### PR TITLE
[Kernel] Add torch custom op for all_reduce

### PR DESCRIPTION
**Pull request by SageMoore**
This PR encapsulates all_reduce inside of a torch custom op. This allows torch.dynamo to run on graphs containing all_reduce.

The current implementation of all_reduce mixes in-place and out-of-place kernels for the sake having a unified interface. Unfortunately, this won't work with torch compile. The problem with the current implementation is that the output tensor can either share a buffer with the input tensor or have its own buffer. This inconsistency is not supported in torch compile. To get around this I've added in-place and out-of-place all_reduce apis to the GroupCoordinator. This allows us to have a custom op for each one and to do the dispatching outside of the ops.

I had some difficulty getting the should_custom_ar kernel working with dynamo. The issue seemed to be centered around its non-tensor return type. For the sake of expediency I just moved that kernel to python as a workaround since it doesn't have any users outside of CustomAllReduce. I'm happy to look into this more if others feel strongly that it should remain in cuda.

### CODE REVIEW

**Comment by youkaichao** [comment]
will take a look later!

File: vllm/distributed/communication_op.py [hunk 1]
```diff
@@ -5,10 +5,41 @@
 
 from .parallel_state import get_tp_group
 
+torch.library.define("vllm::tp_out_of_place_ar",
+                     ("(Tensor(a!) input_ ) -> Tensor"))
 
-def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
+
+@torch.library.register_kernel("vllm::tp_out_of_place_ar", ("cuda", "cpu"))
```
Comment (bnellnm) [inline vllm/distributed/communication_op.py:12]: These can probably be combined to use `torch.library.custom_op`, e.g.

```
@torch.library.custom_op(mutates_args=["input_"], device_types=("cuda", "cpu"))
def _tp_out_of_place_ar ...
```

similarly for the rest.

File: vllm/distributed/communication_op.py [hunk 2]
```diff
@@ -6,9 +6,38 @@
...
+                         device_types=("cuda", "cpu"))
+def _tp_in_place_ar(input_: torch.Tensor) -> None:
     """All-reduce the input tensor across model parallel group."""
-    return get_tp_group().all_reduce(input_)
+    get_tp_group().in_place_ar(input_)
+
+
+@torch.library.register_fake("vllm::tp_in_place_ar")
+def _tp_in_place_ar_fake(input_: torch.Tensor) -> None:
+    return
```
Comment (bnellnm) [inline vllm/distributed/communication_op.py:32]: I don't think nullary ops require meta-functions.

File: vllm/distributed/parallel_state.py [hunk 3]
```diff
@@ -291,7 +287,33 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
...
+            assert tpu_comm is not None
+            return tpu_comm.all_reduce(input_).clone()
+
+        # Otherwise, use the custom kernel
+        assert self.should_use_ca_comm(input_)
+        ca_comm = self.ca_comm
+        assert ca_comm is not None
+        out = ca_comm.custom_all_reduce(input_)
+        assert out is not None
+        return out.clone()
```
Comment (bnellnm) [inline vllm/distributed/parallel_state.py:307]: What happens if out is not cloned here?

File: vllm/distributed/communication_op.py (hunk 2 above)
Comment (SageMoore) [reply vllm/distributed/communication_op.py:32]: Yeah, it looks like that's true.

File: vllm/distributed/parallel_state.py [hunk 4]
```diff
@@ -291,7 +288,33 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
             ipex.distributed.all_reduce(input_, group=self.device_group)
         else:
             torch.distributed.all_reduce(input_, group=self.device_group)
-        return input_
+
+    def out_of_place_ar(self, input_: torch.Tensor) -> torch.Tensor:
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:292]: where do you call this function?

File: vllm/distributed/communication_op.py [hunk 5]
```diff
@@ -6,9 +6,33 @@
...
+@torch.library.custom_op("vllm::tp_in_place_ar",
+                         mutates_args=["input_"],
+                         device_types=("cuda", "cpu"))
+def _tp_in_place_ar(input_: torch.Tensor) -> None:
+    """All-reduce the input tensor across model parallel group."""
+    get_tp_group().in_place_ar(input_)
+
+
+def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
+    if get_tp_group().should_run_out_of_place_ar(input_):
```
Comment (youkaichao) [inline vllm/distributed/communication_op.py:31]: how does Dynamo work with this condition?

**Review by youkaichao — COMMENTED** [review]
I think moving `should_custom_ar` to python makes sense to me. But I need to know more about how Dynamo treats the selection logic you use currently.

sorry for the long wait!

File: vllm/distributed/communication_op.py (hunk 5 above)
Comment (SageMoore) [reply vllm/distributed/communication_op.py:31]: I tested this by running the `pytest tests/distributed/test_basic_distributed_correctness.py` test with dynamo enabled. Everything runs to completion without any errors when fullgraph is True or False.

Is there something specific you are worried about? I'm also happy to run additional tests.

File: vllm/distributed/parallel_state.py (hunk 4 above)
Comment (SageMoore) [reply vllm/distributed/parallel_state.py:292]: https://github.com/vllm-project/vllm/pull/7755/files#diff-4f562f5059f3e90ace1659583afc6fff378aa8db722d8c61995caef06df94cdaR14

**Comment by SageMoore** [comment]
> I think moving `should_custom_ar` to python makes sense to me. But I need to know more about how Dynamo treats the selection logic you use currently.
> 
> sorry for the long wait!

No worries. Thanks for taking a look :)

File: vllm/distributed/communication_op.py (hunk 5 above)
Comment (youkaichao) [reply vllm/distributed/communication_op.py:31]: can you try to follow https://github.com/vllm-project/vllm/blob/baa5467547a758af35f442af6edfbc0fb73c83ce/tests/tpu/test_compilation.py#L13 and see what's the transformed bytecode?

Comment (youkaichao) [reply vllm/distributed/communication_op.py:31]: okay, I think all these conditions turn into guards.

**Comment by youkaichao** [comment]
close as it has been reworked in https://github.com/vllm-project/vllm/pull/8526 . thanks for your pioneering work! @SageMoore 🙏

**Pull request closed without merging**